pages/en/zkapps/simple-anonymous-message-board-tutorial.mdx

```diff
@@ -0,0 +1,345 @@
+
+### Computing `signerPublicKey` from `signerPrivateKey`
+
+Now that we have the user’s private key, we’ll need to derive the associated public key to check it against the list of approved publishers. Luckily the `PrivateKey` type in SnarkyJS includes a `toPublicKey()` method.
+
+```ts
+// Compute signerPublicKey from signerPrivateKey argument
+const signerPublicKey = signerPrivateKey.toPublicKey();
+```
+
+We’ll have to check if this public key matches one of the ones stored on-chain. So let’s grab those as well.
+
+```ts
+// Get approved public keys
+const user1 = this.user1.get();
+const user2 = this.user2.get();
+const user3 = this.user3.get();
+```
+
+Calling the `get()` method tells SnarkyJS to retrieve these values from the zkApp account’s on-chain state. (Note that SnarkyJS uses a single network request to retrieve all on-chain state values simultaneously.)
+
+Finally, we check if `signerPublicKey` is equal to one of the allowed public keys contained in our `user` variables.
+
+```ts
+// Assert that signerPublicKey is one of the approved public keys
+signerPublicKey
+  .equals(user1)
+  .or(signerPublicKey.equals(user2))
+  .or(signerPublicKey.equals(user3))
+  .assertEquals(true);
+```
+
+Notice that we call the SnarkyJS `equals()` and `or()` methods instead of using the JavaScript operators (`===`, and `||`). The built-in SnarkyJS methods have the same effect, but they work with SnarkyJS types, and their execution can be verified using a zero-knowledge proof.
+
+`assertEquals(true)` at the end means that it will be impossible to generate a valid proof unless `signerPublicKey` is equal to one of the pre-approved users. The Mina network will reject any transaction sent to a zkApp account that doesn’t include a valid zero-knowledge proof for that account. So it will be impossible for users to post new messages unless they have a private key associated with one of the three pre-approved public keys.
+
+### Updating `message`
+
+Until now, we have worked to ensure that only approved users can call `publishMessage()`. When they do, the contract should update the on-chain `message` variable to their new message.
+
+```ts
+// Update on-chain message variable
+this.message.set(message);
+```
+
+The `set()` method will ask the Mina nodes to update the value of an on-chain variable, but only if the associated proof is valid.
+
+### Updating `messageHistoryHash`
+
+There’s one more thing we should do. If we want users to be able to keep track of what has been said, then we need to store a commitment to the message history on-chain. There are a few ways to do this, but the simplest is to store a hash of our new `message` and our old `messageHistoryHash` every time we call `publishMessage`.
+
+```ts
+// Computer new messageHistoryHash
+const oldHash = this.messageHistoryHash.get();
+const newHash = Poseidon.hash([message, oldHash]);
+
+// Update on-chain messageHistoryHash
```

Comment:
How about `// Update on-chain state`, for consistency and to reduce change of anybody misinterpreting that messageHistoryHash is a 'special' name